// @react-native-firebase/app/dist/module/types/app.js
"use strict";

/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
/**
 * Core React Native Firebase package types.
 *
 * @firebase app
 */
// eslint-disable-next-line @typescript-eslint/no-namespace
export let ReactNativeFirebase;
(function (_ReactNativeFirebase) {
  /**
   * Base interface for FirebaseApp containing core properties and methods.
   * The concrete FirebaseApp class implements this interface.
   * Module-specific methods (auth(), analytics(), etc.) are added to FirebaseApp via declaration merging.
   */

  /**
   * Full FirebaseApp interface that extends the base interface.
   * Module-specific methods (auth(), analytics(), etc.) are added here via declaration merging
   * from individual package .d.ts files.
   */
  // eslint-disable-next-line @typescript-eslint/no-empty-object-type

  /**
   * Interface for a supplied `AsyncStorage`.
   */

  /**
   * A class that all React Native Firebase modules extend from to provide default behaviour.
   */
  class FirebaseModule {}
  _ReactNativeFirebase.FirebaseModule = FirebaseModule; // eslint-disable-next-line @typescript-eslint/no-empty-object-type
  // eslint-disable-next-line @typescript-eslint/no-empty-object-type
  /**
   * Type for the `firebase` named export from module packages.
   * Provides complete typing for:
   * - Root level access: `firebase.functions(app?)`
   * - App level access: `firebase.app().functions(region?)`
   * - Statics: `firebase.functions.HttpsErrorCode`
   * - Root properties: `firebase.SDK_VERSION`, `firebase.app()`, etc.
   *
   * @typeParam Namespace - The module namespace (e.g., 'functions', 'auth', 'firestore')
   * @typeParam M - The module instance type (must extend FirebaseModule with `app` property)
   * @typeParam S - The module statics type
   * @typeParam HasCustomArg - true if app-level accessor takes optional string (region/url/databaseId)
   *
   * @example
   * // In functions package:
   * export const firebase = getFirebaseRoot() as ReactNativeFirebase.FirebaseNamespacedExport<
   *   'functions',
   *   FunctionsModule,
   *   FunctionsStatics,
   *   true  // functions() takes regionOrCustomDomain
   * >;
   */
})(ReactNativeFirebase || (ReactNativeFirebase = {}));
/**
 * @firebase utils
 */
// eslint-disable-next-line @typescript-eslint/no-namespace
export let Utils;
(function (_Utils) {
  var FirebaseModule = ReactNativeFirebase.FirebaseModule;
  /**
   * A collection of native device file paths to aid in the usage of file path based methods.
   *
   * Concatenate a file path with your target file name when using with Storage `putFile` or `writeToFile`.
   *
   * ```js
   * firebase.utils.FilePath;
   * ```
   */
  /**
   * For further information on the status codes available & what they represent, please head over
   * to ConnectionResult documentation:
   * https://developers.google.com/android/reference/com/google/android/gms/common/ConnectionResult
   */
  let PlayServicesAvailabilityStatusCodes = /*#__PURE__*/function (PlayServicesAvailabilityStatusCodes) {
    PlayServicesAvailabilityStatusCodes[PlayServicesAvailabilityStatusCodes["API_UNAVAILABLE"] = 16] = "API_UNAVAILABLE";
    PlayServicesAvailabilityStatusCodes[PlayServicesAvailabilityStatusCodes["CANCELED"] = 13] = "CANCELED";
    PlayServicesAvailabilityStatusCodes[PlayServicesAvailabilityStatusCodes["DEVELOPER_ERROR"] = 10] = "DEVELOPER_ERROR";
    PlayServicesAvailabilityStatusCodes[PlayServicesAvailabilityStatusCodes["DRIVE_EXTERNAL_STORAGE_REQUIRED"] = 1500] = "DRIVE_EXTERNAL_STORAGE_REQUIRED";
    PlayServicesAvailabilityStatusCodes[PlayServicesAvailabilityStatusCodes["INTERNAL_ERROR"] = 8] = "INTERNAL_ERROR";
    PlayServicesAvailabilityStatusCodes[PlayServicesAvailabilityStatusCodes["INTERRUPTED"] = 15] = "INTERRUPTED";
    PlayServicesAvailabilityStatusCodes[PlayServicesAvailabilityStatusCodes["INVALID_ACCOUNT"] = 5] = "INVALID_ACCOUNT";
    PlayServicesAvailabilityStatusCodes[PlayServicesAvailabilityStatusCodes["LICENSE_CHECK_FAILED"] = 11] = "LICENSE_CHECK_FAILED";
    PlayServicesAvailabilityStatusCodes[PlayServicesAvailabilityStatusCodes["NETWORK_ERROR"] = 7] = "NETWORK_ERROR";
    PlayServicesAvailabilityStatusCodes[PlayServicesAvailabilityStatusCodes["RESOLUTION_REQUIRED"] = 6] = "RESOLUTION_REQUIRED";
    PlayServicesAvailabilityStatusCodes[PlayServicesAvailabilityStatusCodes["RESTRICTED_PROFILE"] = 20] = "RESTRICTED_PROFILE";
    PlayServicesAvailabilityStatusCodes[PlayServicesAvailabilityStatusCodes["SERVICE_DISABLED"] = 3] = "SERVICE_DISABLED";
    PlayServicesAvailabilityStatusCodes[PlayServicesAvailabilityStatusCodes["SERVICE_INVALID"] = 9] = "SERVICE_INVALID";
    PlayServicesAvailabilityStatusCodes[PlayServicesAvailabilityStatusCodes["SERVICE_MISSING"] = 1] = "SERVICE_MISSING";
    PlayServicesAvailabilityStatusCodes[PlayServicesAvailabilityStatusCodes["SERVICE_MISSING_PERMISSION"] = 19] = "SERVICE_MISSING_PERMISSION";
    PlayServicesAvailabilityStatusCodes[PlayServicesAvailabilityStatusCodes["SERVICE_UPDATING"] = 18] = "SERVICE_UPDATING";
    PlayServicesAvailabilityStatusCodes[PlayServicesAvailabilityStatusCodes["SERVICE_VERSION_UPDATE_REQUIRED"] = 2] = "SERVICE_VERSION_UPDATE_REQUIRED";
    PlayServicesAvailabilityStatusCodes[PlayServicesAvailabilityStatusCodes["SIGN_IN_FAILED"] = 17] = "SIGN_IN_FAILED";
    PlayServicesAvailabilityStatusCodes[PlayServicesAvailabilityStatusCodes["SIGN_IN_REQUIRED"] = 4] = "SIGN_IN_REQUIRED";
    PlayServicesAvailabilityStatusCodes[PlayServicesAvailabilityStatusCodes["SUCCESS"] = 0] = "SUCCESS";
    PlayServicesAvailabilityStatusCodes[PlayServicesAvailabilityStatusCodes["TIMEOUT"] = 14] = "TIMEOUT";
    return PlayServicesAvailabilityStatusCodes;
  }({});
  _Utils.PlayServicesAvailabilityStatusCodes = PlayServicesAvailabilityStatusCodes;
  /**
   * The React Native Firebase Utils service interface.
   *
   * > This module is available for the default app only.
   *
   * #### Example
   *
   * Get the Utils service for the default app:
   *
   * ```js
   * const defaultAppUtils = firebase.utils();
   * ```
   */
  class Module extends FirebaseModule {}
  _Utils.Module = Module;
})(Utils || (Utils = {})); // Re-export FirebaseApp as a named type for easier importing
